cmd/soroban-cli/src/commands/contract/build.rs

```diff
@@ -0,0 +1,189 @@
+use clap::Parser;
+use itertools::Itertools;
+use std::{
+    collections::HashSet,
+    ffi::OsStr,
+    fmt::Debug,
+    fs, io,
+    path::Path,
+    process::{Command, ExitStatus, Stdio},
+};
+
+use cargo_metadata::{Metadata, MetadataCommand, Package};
+
+/// Build a contract from source
+///
+/// Builds all crates that are referenced by the cargo manifest (Cargo.toml)
+/// that have cdylib as their crate-type. Crates are built for the wasm32
+/// target. Unless configured otherwise, crates are built with their default
+/// features and with their release profile.
+///
+/// To view the commands that will be executed, without executing them, use the
+/// --print-commands-only option.
+#[derive(Parser, Debug, Clone)]
+pub struct Cmd {
+    /// Path to Cargo.toml
+    #[arg(long, default_value = "Cargo.toml")]
+    pub manifest_path: std::path::PathBuf,
+    /// Package to build
+    ///
+    /// If omitted, all packages that build for crate-type cdylib are built.
+    #[arg(long)]
+    pub package: Option<String>,
+    /// Build with the specified profile
+    #[arg(long, default_value = "release")]
+    pub profile: String,
+    /// Build with the list of features activated, space or comma separated
+    #[arg(long)]
+    pub features: Option<String>,
+    /// Build with the all features activated
+    #[arg(
+        long,
+        conflicts_with = "features",
+        conflicts_with = "no_default_features"
+    )]
+    pub all_features: bool,
+    /// Build with the default feature not activated
+    #[arg(long)]
+    pub no_default_features: bool,
+    /// Directory to copy wasm files to
+    ///
+    /// If provided, wasm files can be found in the cargo target directory, and
+    /// the specified directory.
+    ///
+    /// If ommitted, wasm files are written only to the cargo target directory.
+    #[arg(long)]
+    pub out_dir: Option<std::path::PathBuf>,
+    /// Print commands to build without executing them
+    #[arg(long)]
+    pub print_commands_only: bool,
```

Comment:
dry run ?

Reply:
I'm contemplating removing the feature, to keep the feature set a little slimmer. Also, because not everything this command does is captured in cargo commands as output. Let me know if you think we should keep it.

Reply:
Actually, the option is really helpful for testing and development.

Reply:
rehashing /w myself, we're not the first ones to invoke a compiler under the hood. Both rustc and go are invoking the gnu compiler, and both have options to allow the enduser to pass in arbitrary parameters and/or extra printing.

Reply:
I named it `--print-commands-only` because that's explicitly describing what it does. The name of the option is similar to the same option that's provided in the `cargo-hack` tool.

I can rename it to dry run if that is preferred / makes more sense.